File: backend/src/main/java/com/back/fairytale/global/security/CustomOAuth2User.java
package com.back.fairytale.global.security;

import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.oauth2.core.user.DefaultOAuth2User;
import org.springframework.security.oauth2.core.user.OAuth2User;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;

@Getter
@RequiredArgsConstructor
public class CustomOAuth2User implements OAuth2User {

    private final Long id;
    private final String username;
    private final String role;

    @Override
    public String getName() {
        return username;
    }

    @Override
    public Map<String, Object> getAttributes() {
        return Map.of();
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        List<GrantedAuthority> collection = new ArrayList<>();
        collection.add((GrantedAuthority) () -> role);
        return collection;
    }
}

File: backend/src/main/java/com/back/fairytale/global/security/CustomOAuth2User.kt
package com.back.fairytale.global.security

import org.springframework.security.core.GrantedAuthority
import org.springframework.security.oauth2.core.user.OAuth2User
import java.util.Map

class CustomOAuth2User(
    val id: Long,
    val username: String,
    val role: String
) : OAuth2User {

    override fun getName(): String = username

    override fun getAttributes(): MutableMap<String, Any> = mutableMapOf()

    override fun getAuthorities(): Collection<GrantedAuthority> {
        return listOf(GrantedAuthority { role })
    }
}
